[GeoMechanicsApplication] Update the piping analysis method to allow for sloped pipe trajectories

mcgicjn2 opened the pull request:
**📝 Description**
Updated the piping slope to allow for sloping pipes.

**🆕 Changelog**
- Changed from water_pressure based to hydraulic_head based calculation across pipes
- Added pipe slope (i.e. calculated not 0 )
- updated some * to ::pointer
- created fixed cpp_unit_test for erosion (original files are changed by python tests)

mcgicjn2 committed:
Initial Commit of Sloping Pipe Elements

This is a first commit of updating the pipe elements to incorporate angles. Tests are still needed

mcgicjn2 committed:
Fixed failing test on Github Tests

- The test element was set to 1m, while there was an attempt to override the pipe_length to 0.5m, removed override, and made test element 0.5m
- Adjusted CalculateWaterPressure to only take the requried parameter (Geom)

mcgicjn2 committed:
Updated Piping Element to use head gradient in equilibrium method

mcgicjn2 committed:
Updated element length

mcgicjn2 committed:
Updated Aron's equilibrium piping equation to account for heads and not pressures

mcgicjn2 committed:
Tidied up code

mcgicjn2 committed:
Created fixed test file for running cpp_unit_tests from piping erosion stategy

aronnoordam commented on applications/GeoMechanicsApplication/custom_elements/steady_state_Pw_piping_element.cpp:
```diff
@@ -96,25 +97,61 @@ Initialize(const ProcessInfo& rCurrentProcessInfo)
+
+            const double WaterPressure = rGeom[node].FastGetSolutionStepValue(WATER_PRESSURE);
+            NodalHydraulicHead[node] = -inner_prod(NodeCoordinates, NodeVolumeAccelerationUnitVector)
+                - PORE_PRESSURE_SIGN_FACTOR * WaterPressure / FluidWeight;
```
I dont think this is correct above the saturated zone, in the case where the retention model is chosen such that above the saturated zone, the soil is assumed to be unsaturated

aronnoordam replied:
since, above the saturated zone, the Water pressure is 0. Given this formulation, the head is now defined as the vertical coordinate

aronnoordam replied:
actually, setting the water pressure to 0 above the phreatic line is not implemented. So for now this code works. but we might have to look at it in the future

aronnoordam commented on applications/GeoMechanicsApplication/custom_elements/steady_state_Pw_piping_element.cpp:
```diff
@@ -306,52 +366,35 @@ double SteadyStatePwPipingElement<TDim, TNumNodes>::CalculateParticleDiameter(co
-	
-    return modelFactor * M_PI / 3.0 * particle_d * (SolidDensity - FluidDensity) * gravity * eta  * sin((theta  + pipeSlope) * M_PI / 180.0) / cos(theta * M_PI / 180.0) / dpdx;
-
+    return modelFactor * M_PI / 3.0 * particle_d * (SolidDensity - FluidDensity) * gravity * eta * sin((theta + pipeSlope) * M_PI / 180.0) / cos((theta * M_PI / 180.0)) / (dhdl * FluidDensity * gravity);
```
gravity can be removed from the equation, as you multiply and divide by the gravity

aronnoordam commented on applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp:
```diff
@@ -388,40 +389,40 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy :
                     // get open pipe element geometry and properties
-                    auto& Geom = OpenPipeElement->GetGeometry();
-                    auto& prop = OpenPipeElement->GetProperties();
+                    auto& pGeom = pElement->GetGeometry();
```
rGeom and rProp, as you are using references. pGeom and pProp indicate that Geom and Prop are refered to as pointers according to the kratos styling documentation

mcgicjn2 committed:
Reduced code smells

mcgicjn2 committed:
Updated typo in logical condition

mcgicjn2 committed:
Reducing code smells further + improve variable name

mcgicjn2 committed:
Updated misnamed variable

mcgicjn2 committed:
Made changes adding & to smart pointers

mcgicjn2 committed:
Reduced code smells further based on SonarQube

mcgicjn2 committed:
removed new bugs and code smells

carloslubbers committed:
Merge branch 'master' into geo/10641-piping-slope

carloslubbers committed:
Merge branch 'master' into geo/10641-piping-slope

carloslubbers committed:
Merge branch 'master' into geo/10641-piping-slope

mcgicjn2 committed:
Updated Piping Investigation

Changed the equilibrium pipe height equation to align with that of JvE.

Also refactored and seperated the length and slope calculations.
